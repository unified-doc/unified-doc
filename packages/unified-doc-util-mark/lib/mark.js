import textOffsets from 'unified-doc-util-text-offsets';
import visit from 'unist-util-visit';

import {
  getMarkedNodes,
  getOverLappingMarks,
  getNodeSegments,
} from './mark-node';
import validateMarks from './validate-marks';

export default function mark(hast, marks = []) {
  if (marks.length === 0) {
    return hast;
  }

  const hastWithOffsets = textOffsets(hast);
  const validatedMarks = validateMarks(marks);

  visit(hastWithOffsets, 'text', (node, _index, parent) => {
    if (!node.data || !node.data.textOffset) {
      return visit.CONTINUE;
    }
    const overlappingMarks = getOverLappingMarks(node, validatedMarks);
    if (overlappingMarks.length === 0) {
      return visit.CONTINUE;
    }
    const appliedMarkIds = new Set();
    const nodeSegments = getNodeSegments(node, overlappingMarks);
    const markedNodes = getMarkedNodes(nodeSegments, appliedMarkIds);

    // Reconstruct nodes under parent
    const siblings = parent.children;
    const currentNodeIndex = siblings.indexOf(node);
    parent.children = siblings
      .slice(0, currentNodeIndex)
      .concat(markedNodes)
      .concat(siblings.slice(currentNodeIndex + 1));
    return visit.CONTINUE;
  });

  return hastWithOffsets;
}
