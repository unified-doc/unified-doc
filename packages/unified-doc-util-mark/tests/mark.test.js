import { hast } from './fixtures';
import mark from '../lib/mark';

describe(mark, () => {
  it('returns the original tree if there is nothing to mark', () => {
    expect(mark(hast)).toBe(hast);
  });

  it('returns a new tree if there are no matching marks', () => {
    const marked = mark(hast, [{ id: 'a', start: 1000, end: 4000 }]);
    expect(marked).toEqual(hast);
    expect(marked).not.toBe(hast);
  });

  it('marks non-overlapping text nodes', () => {
    const marked = mark(hast, [{ id: 'a', start: 7, end: 15 }]);
    expect(marked).not.toEqual(hast);
    expect(marked).toMatchSnapshot();
  });

  it('marks overlapping text nodes', () => {
    const marked = mark(hast, [
      { id: 'a', start: 7, end: 15 },
      { id: 'b', start: 10, end: 20 },
    ]);
    expect(marked).not.toEqual(hast);
    expect(marked).toMatchSnapshot();
  });
});
