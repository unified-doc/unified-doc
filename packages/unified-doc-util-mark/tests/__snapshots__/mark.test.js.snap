// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mark marks non-overlapping text nodes 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "data": Object {
            "textOffset": Object {
              "end": 1,
              "start": 0,
            },
          },
          "type": "text",
          "value": "
",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "data": Object {
                    "textOffset": Object {
                      "end": 5,
                      "start": 1,
                    },
                  },
                  "position": Object {
                    "end": Object {
                      "column": 9,
                      "line": 1,
                      "offset": 8,
                    },
                    "start": Object {
                      "column": 5,
                      "line": 1,
                      "offset": 4,
                    },
                  },
                  "type": "text",
                  "value": "some",
                },
              ],
              "position": Object {
                "end": Object {
                  "column": 11,
                  "line": 1,
                  "offset": 10,
                },
                "start": Object {
                  "column": 3,
                  "line": 1,
                  "offset": 2,
                },
              },
              "properties": Object {},
              "tagName": "strong",
              "type": "element",
            },
            Object {
              "type": "text",
              "value": " m",
            },
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "arkdown ",
                },
              ],
              "properties": Object {
                "dataMarkId": "a",
                "id": "a",
              },
              "tagName": "mark",
              "type": "element",
            },
            Object {
              "type": "text",
              "value": "content",
            },
          ],
          "position": Object {
            "end": Object {
              "column": 28,
              "line": 1,
              "offset": 27,
            },
            "start": Object {
              "column": 3,
              "line": 1,
              "offset": 2,
            },
          },
          "properties": Object {},
          "tagName": "p",
          "type": "element",
        },
        Object {
          "data": Object {
            "textOffset": Object {
              "end": 23,
              "start": 22,
            },
          },
          "type": "text",
          "value": "
",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 28,
          "line": 1,
          "offset": 27,
        },
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "properties": Object {},
      "tagName": "blockquote",
      "type": "element",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 28,
      "line": 1,
      "offset": 27,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`mark marks overlapping text nodes 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "data": Object {
            "textOffset": Object {
              "end": 1,
              "start": 0,
            },
          },
          "type": "text",
          "value": "
",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "data": Object {
                    "textOffset": Object {
                      "end": 5,
                      "start": 1,
                    },
                  },
                  "position": Object {
                    "end": Object {
                      "column": 9,
                      "line": 1,
                      "offset": 8,
                    },
                    "start": Object {
                      "column": 5,
                      "line": 1,
                      "offset": 4,
                    },
                  },
                  "type": "text",
                  "value": "some",
                },
              ],
              "position": Object {
                "end": Object {
                  "column": 11,
                  "line": 1,
                  "offset": 10,
                },
                "start": Object {
                  "column": 3,
                  "line": 1,
                  "offset": 2,
                },
              },
              "properties": Object {},
              "tagName": "strong",
              "type": "element",
            },
            Object {
              "type": "text",
              "value": " m",
            },
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "ark",
                },
              ],
              "properties": Object {
                "dataMarkId": "a",
                "id": "a",
              },
              "tagName": "mark",
              "type": "element",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [
                    Object {
                      "type": "text",
                      "value": "down ",
                    },
                  ],
                  "properties": Object {
                    "dataMarkId": "b",
                    "id": "b",
                  },
                  "tagName": "mark",
                  "type": "element",
                },
              ],
              "properties": Object {
                "dataMarkId": "a",
              },
              "tagName": "mark",
              "type": "element",
            },
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "conte",
                },
              ],
              "properties": Object {
                "dataMarkId": "b",
              },
              "tagName": "mark",
              "type": "element",
            },
            Object {
              "type": "text",
              "value": "nt",
            },
          ],
          "position": Object {
            "end": Object {
              "column": 28,
              "line": 1,
              "offset": 27,
            },
            "start": Object {
              "column": 3,
              "line": 1,
              "offset": 2,
            },
          },
          "properties": Object {},
          "tagName": "p",
          "type": "element",
        },
        Object {
          "data": Object {
            "textOffset": Object {
              "end": 23,
              "start": 22,
            },
          },
          "type": "text",
          "value": "
",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 28,
          "line": 1,
          "offset": 27,
        },
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "properties": Object {},
      "tagName": "blockquote",
      "type": "element",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 28,
      "line": 1,
      "offset": 27,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;
